src/runtime/virtcontainers/utils/utils.go

```diff
@@ -306,6 +306,36 @@ func ConvertAddressFamily(family int32) pbTypes.IPFamily {
 	}
 }
 
+func waitProcessUsingWaitLoop(pid int, timeoutSecs uint, logger *logrus.Entry) bool {
+	secs := time.Duration(timeoutSecs)
+	timeout := time.After(secs * time.Second)
+
+	for {
+		// Check if the process is running periodically to avoid a busy loop
+		// "A watched pot never boils" and an unwaited-for process never appears to die!
```

Comment:
> I don't understand that comment at all, nor the logic in the loop (which appears really similar to what was there before). 

This is code refactoring. The comment and the logic mostly come from the existing code base.

> Is it saying that if you don't have a `wait4`, then `kill(pid, 0)` always appear successful?

Not exactly. It says that `kill(pid, 0)` will succeed until the process is reaped by its parent.

The confusion in this code comes from the fact it must support the hypervisor process being a direct child or not. The latter case, e.g. QEMU with the current code base, can only be achieved with `kill(0, pid)` when you don't have pidfds.
The direct child case cannot be handled by `kill(0, pid)` alone : something must reap the child first.

I don't like the comment either because if seems to indicate that `wait4` is there _just_ to ensure `kill(0)` will return ESRCH... while the important aspect of `wait4` is primarily to reap dead child resources.

Also this `kill(0)` construct is fragile as we have no guarantees that pid hasn't been recycled : things would be much more sane if hypervisor processses were always direct children because the pid is guaranteed to be valid until waited for.

Reply:
I think the original intent was for wait4 to handle the case for child processes and `kill(0, pid)` for the non-child processes. I removed the comment, it is confusing.

Reply:
@gkurz I like your explanation, it is clear and makes for a nice addition to the comment.
With that addition, we could keep "A watched pot never boils" comment, in case it's a "mind trigger" for the original authors. It just needs to be explained.

Reply:
Nit: the comment gets removed by the next patch (runtime: use exponential backoff for process wait) instead of this patch. Also, I'd still leave a comment that explains why we're looping on both kill(0) (non child process) and wait4(WNOHANG) (child process).